kernel/pipes.c

```diff
@@ -557,173 +476,156 @@ static int pipe_get_internal(k_spinlock_key_t key, struct k_pipe *pipe,
 			     size_t *bytes_read, size_t min_xfer,
 			     k_timeout_t timeout)
 {
-	struct k_thread    *writer;
-	struct k_pipe_desc *desc;
-	sys_dlist_t    xfer_list;
-	size_t         num_bytes_read = 0;
-	size_t         data_off;
+	sys_dlist_t         src_list;
+	struct k_pipe_desc  pipe_desc[2];
+	struct k_pipe_desc  dest_desc;
+	struct k_pipe_desc  *src_desc;
+	size_t         num_bytes_read = 0U;
 	size_t         bytes_copied;
+	size_t         bytes_can_read = 0U;
+	bool           reschedule_needed = false;
 
 	/*
-	 * Create a list of "working readers" into which the data will be
-	 * directly copied.
+	 * Data copying takes place in the following order.
+	 * 1. Copy data from the pipe buffer to the receive buffer.
+	 * 2. Copy data from the waiting writer(s) to the receive buffer.
+	 * 3. Refill the pipe buffer from the waiting writer(s).
 	 */
 
-	if (!pipe_xfer_prepare(&xfer_list, &writer, &pipe->wait_q.writers,
-				pipe->bytes_used, bytes_to_read,
-				min_xfer, timeout)) {
+	sys_dlist_init(&src_list);
+
+	if (pipe->bytes_used != 0) {
+		bytes_can_read = pipe_buffer_list_populate(&src_list,
+							   pipe_desc,
+							   pipe->buffer,
+							   pipe->size,
+							   pipe->read_index,
+							   pipe->write_index);
+	}
+
+	bytes_can_read += pipe_waiter_list_populate(&src_list,
+						    &pipe->wait_q.writers,
+						    bytes_to_read);
+
+	if ((bytes_can_read < min_xfer) &&
+	    (K_TIMEOUT_EQ(timeout, K_NO_WAIT))) {
+
+		/* The request can not be fulfilled. */
+
 		k_spin_unlock(&pipe->lock, key);
 		*bytes_read = 0;
 
 		return -EIO;
 	}
 
-	z_sched_lock();
-	k_spin_unlock(&pipe->lock, key);
-
-	num_bytes_read = pipe_buffer_get(pipe, data, bytes_to_read);
+	dest_desc.buffer = data;
+	dest_desc.bytes_to_xfer = bytes_to_read;
+	dest_desc.thread = _current;
 
-	/*
-	 * 1. 'xfer_list' currently contains a list of writer threads that can
-	 *     have their write requests fulfilled by the current call.
-	 * 2. 'writer' if not NULL points to a thread on the writer wait_q
-	 *    that can post some of its requested data.
-	 * 3. Data will be copied from each writer's buffer to either the
-	 *    reader's buffer and/or to the pipe's circular buffer.
-	 * 4. Interrupts are unlocked but the scheduler is locked to allow
-	 *    ticks to be delivered but no scheduling to occur
-	 * 5. If 'writer' times out while we are copying data, not only do we
-	 *    still have a pointer to it, but it can not execute until this
-	 *    call is complete so it is still safe to copy data from it.
-	 */
+	src_desc = (struct k_pipe_desc *)sys_dlist_get(&src_list);
+	while (src_desc != NULL) {
```

Comment:
The logic in this loop is almost but not quite identical to the same copy loop in k_pipe_put(), and I'm wondering if it's worth further unifying the treatment.  Both schemes are just iterating over two lists of buffers to copy N bytes between them, it's just that one side of each version is guaranteed to have only one element (the buffer provided to the API by the caller).

But beyond that?  They're the same code.  And the "only one element" distinction is something that can be handled with just one extra test at the top of the loop (and an extra dummy list head somewhere).

I love the semantic unification of this PR, but it does seem like there's more fruit on the tree.

Reply:
Let's see what I can do. :)

Reply:
On the surface, there are a lot of similarities in the two loops. So much so that it does seem natural enough to want to unify them. However, there are a substantial number of minor differences that stem from the direction that data flows (adding vs subtraction, whether its a source or destination descriptor's thread that must be NULL, ...) . With the current proposal, my initial attempts just generate a hacky mess--nothing elegant (yet).

Consequently, I would like to propose that we track this specific cleanup request as an enhancement.